pallets/gear-voucher/src/lib.rs

```diff
@@ -100,151 +98,413 @@ pub mod pallet {
+        UnknownDestination,
+        /// Voucher has expired and couldn't be used.
+        VoucherExpired,
+        /// Zero validity given while issuing voucher.
+        ZeroValidity,
     }
 
-    #[pallet::hooks]
-    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+    /// Storage containing amount of the total vouchers issued.
+    ///
+    /// Used as nonce in voucher creation.
+    #[pallet::storage]
+    type Issued<T> = StorageValue<_, u64>;
+
+    // Public wrap of the amount of issued vouchers.
+    common::wrap_storage_value!(storage: Issued, name: IssuedWrap, value: u64);
+
+    /// Double map storage containing data of the voucher,
+    /// associated with some spender and voucher ids.
+    #[pallet::storage]
+    pub type Vouchers<T> = StorageDoubleMap<
+        _,
+        Identity,
+        AccountIdOf<T>,
+        Identity,
+        VoucherId,
+        VoucherInfo<AccountIdOf<T>, BlockNumberFor<T>>,
+    >;
 
     #[pallet::call]
-    impl<T: Config> Pallet<T> {
-        /// Issue a new voucher for a `user` to be used to pay for sending messages
-        /// to `program_id` program.
-        ///
-        /// The dispatch origin for this call must be _Signed_.
+    impl<T: Config> Pallet<T>
+    where
+        T::AccountId: Origin,
+    {
+        /// Issue a new voucher.
         ///
-        /// - `to`: The voucher holder account id.
-        /// - `program`: The program id, messages to whom can be paid with the voucher.
-        /// NOTE: the fact a program with such id exists in storage is not checked - it's
-        /// a caller's responsibility to ensure the consistency of the input parameters.
-        /// - `amount`: The voucher amount.
+        /// Deposits event `VoucherIssued`, that contains `VoucherId` to be
+        /// used by spender for balance-less on-chain interactions.
         ///
-        /// ## Complexity
-        /// O(Z + C) where Z is the length of the call and C its execution weight.
+        /// Arguments:
+        /// * spender:  user id that is eligible to use the voucher;
+        /// * balance:  voucher balance could be used for tx fees and gas,
+        /// * programs: pool of programs spender can interact with,
+        ///             if None - means any program,
+        ///             limited by Config param;
+        /// * validity: amount of blocks voucher could be used by spender
+        ///             and couldn't be revoked by owner,
+        ///             couldn't be zero.
         #[pallet::call_index(0)]
-        #[pallet::weight(<T as Config>::WeightInfo::issue())]
+        #[pallet::weight(T::WeightInfo::issue())]
         pub fn issue(
             origin: OriginFor<T>,
-            to: AccountIdLookupOf<T>,
-            program: ProgramId,
-            value: BalanceOf<T>,
+            spender: AccountIdOf<T>,
+            balance: BalanceOf<T>,
+            programs: Option<BTreeSet<ProgramId>>,
+            validity: BlockNumberFor<T>,
```

Comment:
isn't _duration_ a common name for that?

Reply:
> isn't _duration_ a common name for that?

If not _validity_, it'd rather be _expiry_ than _duration_: this value determines until which block a voucher is valid, not for how many blocks it's going to be valid.

Reply:
the exact expiration block number seems more logical for using the voucher